File: C++/Ejercicio10.cpp
#include <iostream>
#include <list>
#include <string>
#include <chrono>
#include <thread>
#include <iomanip>

using namespace std;

// Clase para representar una Persona en la fila
class Persona {
public:
    static int contador;
    int id;
    string nombre;
    string motivo;
    chrono::system_clock::time_point horaLlegada;
    int prioridad; // 1: Alta, 2: Normal, 3: Baja
    
    Persona(string nombre, string motivo, int prioridad) {
        this->id = ++contador;
        this->nombre = nombre;
        this->motivo = motivo;
        this->prioridad = prioridad;
        this->horaLlegada = chrono::system_clock::now();
    }
    
    long getTiempoEspera() const {
        auto ahora = chrono::system_clock::now();
        auto duracion = chrono::duration_cast<chrono::seconds>(ahora - horaLlegada);
        return duracion.count() / 60; // Convertir a minutos
    }
    
    string getPrioridadTexto() const {
        switch (prioridad) {
            case 1: return "🔴 ALTA";
            case 2: return "🟡 NORMAL";
            case 3: return "🟢 BAJA";
            default: return "DESCONOCIDA";
        }
    }
    
    string toString() const {
        return "ID: #" + to_string(id) + " | Nombre: " + nombre + " | Motivo: " + motivo + 
               " | Prioridad: " + getPrioridadTexto() + " | Tiempo espera: " + 
               to_string(getTiempoEspera()) + " min";
    }
};

int Persona::contador = 0;

list<Persona> fila;
list<Persona> atendidas;
int personasAtendidas = 0;

void mostrarMenu() {
    cout << "\n========== SISTEMA DE FILA DE ATENCIÓN ==========" << endl;
    cout << "1. Agregar persona a la fila" << endl;
    cout << "2. Ver fila de espera" << endl;
    cout << "3. Atender siguiente persona" << endl;
    cout << "4. Ver estadísticas" << endl;
    cout << "5. Simular atención rápida (5 personas)" << endl;
    cout << "6. Ver personas ya atendidas" << endl;
    cout << "7. SALIR" << endl;
    cout << "=================================================" << endl;
}

// Agregar una persona a la fila
void agregarPersona() {
    cout << "\nIngresa el nombre: ";
    string nombre;
    getline(cin, nombre);
    
    cout << "Ingresa el motivo: ";
    string motivo;
    getline(cin, motivo);
    
    cout << "\nNivel de prioridad:" << endl;
    cout << "  1 - Urgencia (Prioridad ALTA)" << endl;
    cout << "  2 - Normal (Prioridad NORMAL)" << endl;
    cout << "  3 - Consulta (Prioridad BAJA)" << endl;
    cout << "Selecciona (1-3): ";
    int prioridad;
    cin >> prioridad;
    cin.ignore();
    
    if (prioridad < 1 || prioridad > 3) {
        cout << "❌ Prioridad inválida. Se asignó NORMAL." << endl;
        prioridad = 2;
    }
    
    Persona nueva(nombre, motivo, prioridad);
    
    // Insertar respetando prioridades
    if (prioridad == 1) {
        // Alta prioridad: ir al principio de otras altas
        auto it = fila.begin();
        for (; it != fila.end(); ++it) {
            if (it->prioridad > 1) break;
        }
        fila.insert(it, nueva);
    } else {
        // Normal o baja: ir al final
        fila.push_back(nueva);
    }
    
    cout << "✓ " << nombre << " agregado a la fila." << endl;
    
    // Obtener posición
    int posicion = 1;
    for (const auto& p : fila) {
        if (p.id == nueva.id) {
            cout << "  Posición en la fila: #" << posicion << endl;
            break;
        }
        posicion++;
    }
}

// Ver fila de espera
void verFila() {
    if (fila.empty()) {
        cout << "\n✓ La fila está vacía. No hay personas esperando." << endl;
        return;
    }
    
    cout << "\n========== FILA DE ESPERA ==========" << endl;
    cout << "Total personas en fila: " << fila.size() << "\n" << endl;
    
    int contador = 0;
    for (const Persona& p : fila) {
        string marca = (contador == 0) ? "👤 [ATENDIENDO] " : "  [" + to_string(contador + 1) + "] ";
        cout << marca << p.toString() << endl;
        contador++;
    }
}

// Atender a la siguiente persona
void atenderPersona() {
    if (fila.empty()) {
        cout << "\n✓ No hay personas en la fila." << endl;
        return;
    }
    
    Persona persona = fila.front();
    fila.pop_front();
    long tiempoEspera = persona.getTiempoEspera();
    personasAtendidas++;
    
    cout << "\n========== ATENDIENDO ==========" << endl;
    cout << "✓ " << persona.nombre << endl;
    cout << "  ID: #" << persona.id << endl;
    cout << "  Motivo: " << persona.motivo << endl;
    cout << "  Prioridad: " << persona.getPrioridadTexto() << endl;
    cout << "  Tiempo de espera: " << tiempoEspera << " minutos" << endl;
    
    cout << "\n¿Cuánto tiempo tardará la atención? (segundos): ";
    int segundos;
    cin >> segundos;
    cin.ignore();
    
    cout << "⏳ Atendiendo a " << persona.nombre << "..." << endl;
    
    // Simular tiempo de atención
    int tiempoSimulacion = min(segundos / 10, 3); // Máximo 3 segundos
    this_thread::sleep_for(chrono::milliseconds(tiempoSimulacion * 100));
    
    atendidas.push_back(persona);
    cout << "✓ Atención completada." << endl;
    cout << "  Personas en la fila: " << fila.size() << endl;
}

// Ver estadísticas
void verEstadisticas() {
    cout << "\n========== ESTADÍSTICAS ==========" << endl;
    cout << "Personas en fila: " << fila.size() << endl;
    cout << "Personas atendidas: " << personasAtendidas << endl;
    cout << "Total procesadas: " << (fila.size() + personasAtendidas) << endl;
    
    if (!fila.empty()) {
        cout << "\n✓ Próxima persona a atender:" << endl;
        cout << "  " << fila.front().toString() << endl;
        
        cout << "\n✓ Última persona en fila:" << endl;
        cout << "  " << fila.back().toString() << endl;
        
        // Estadísticas por prioridad
        int alta = 0, normal = 0, baja = 0;
        long tiempoPromedioEspera = 0;
        
        for (const Persona& p : fila) {
            if (p.prioridad == 1) alta++;
            else if (p.prioridad == 2) normal++;
            else baja++;
            tiempoPromedioEspera += p.getTiempoEspera();
        }
        
        cout << "\n✓ Distribución por prioridad:" << endl;
        cout << "  🔴 Urgencia: " << alta << endl;
        cout << "  🟡 Normal: " << normal << endl;
        cout << "  🟢 Baja: " << baja << endl;
        
        if (fila.size() > 0) {
            cout << fixed << setprecision(1) 
                 << "  Tiempo promedio espera: " 
                 << (double)tiempoPromedioEspera / fila.size() << " min" << endl;
        }
    }
}

// Simular atención rápida
void simularAtencion() {
    if (fila.empty()) {
        cout << "\n✓ No hay personas en la fila para atender." << endl;
        return;
    }
    
    int atender = min(5, (int)fila.size());
    cout << "\n⏳ Simulando atención a " << atender << " personas...\n" << endl;
    
    for (int i = 0; i < atender; i++) {
        if (!fila.empty()) {
            Persona persona = fila.front();
            fila.pop_front();
            atendidas.push_back(persona);
            personasAtendidas++;
            cout << (i + 1) << ". ✓ " << persona.nombre << " - Atendido" << endl;
            
            this_thread::sleep_for(chrono::milliseconds(500));
        }
    }
    
    cout << "\n✓ Simulación completada." << endl;
    cout << "Personas en fila: " << fila.size() << endl;
    cout << "Personas atendidas: " << personasAtendidas << endl;
}

// Ver personas ya atendidas
void verPersonasAtendidas() {
    if (atendidas.empty()) {
        cout << "\n✓ No hay personas atendidas aún." << endl;
        return;
    }
    
    cout << "\n========== PERSONAS ATENDIDAS ==========" << endl;
    cout << "Total: " << atendidas.size() << "\n" << endl;
    
    int contador = 1;
    for (const Persona& p : atendidas) {
        cout << contador << ". " << p.toString() << endl;
        contador++;
    }
}

int main() {
    int opcion;
    
    // Agregar algunas personas de ejemplo
    fila.push_back(Persona("Ana Martínez", "Consulta", 2));
    fila.push_back(Persona("Carlos López", "Trámite", 3));
    fila.push_back(Persona("Rosa García", "Urgencia", 1));
    
    do {
        mostrarMenu();
        cout << "Selecciona una opción: ";
        cin >> opcion;
        cin.ignore();
        
        switch (opcion) {
            case 1:
                agregarPersona();
                break;
            case 2:
                verFila();
                break;
            case 3:
                atenderPersona();
                break;
            case 4:
                verEstadisticas();
                break;
            case 5:
                simularAtencion();
                break;
            case 6:
                verPersonasAtendidas();
                break;
            case 7:
                cout << "\n¡Gracias por usar el sistema!" << endl;
                break;
            default:
                cout << "\n❌ Opción inválida.\n" << endl;
        }
    } while (opcion != 7);
    
    return 0;
}

File: Java/Ejercicio10.java
import java.util.LinkedList;
import java.util.Scanner;
import java.time.LocalTime;
import java.time.temporal.ChronoUnit;

public class Ejercicio10 {
    
    // Clase interna para representar una Persona en la fila
    static class Persona {
        static int contador = 0;
        int id;
        String nombre;
        String motivo;
        LocalTime horaLlegada;
        int prioridad; // 1: Alta, 2: Normal, 3: Baja
        
        Persona(String nombre, String motivo, int prioridad) {
            this.id = ++contador;
            this.nombre = nombre;
            this.motivo = motivo;
            this.prioridad = prioridad;
            this.horaLlegada = LocalTime.now();
        }
        
        long getTiempoEspera() {
            return ChronoUnit.MINUTES.between(horaLlegada, LocalTime.now());
        }
        
        String getPrioridadTexto() {
            switch (prioridad) {
                case 1: return "🔴 ALTA";
                case 2: return "🟡 NORMAL";
                case 3: return "🟢 BAJA";
                default: return "DESCONOCIDA";
            }
        }
        
        public String toString() {
            return "ID: #" + id + " | Nombre: " + nombre + " | Motivo: " + motivo + 
                   " | Prioridad: " + getPrioridadTexto() + " | Tiempo espera: " + getTiempoEspera() + " min";
        }
    }
    
    static LinkedList<Persona> fila = new LinkedList<>();
    static LinkedList<Persona> atendidas = new LinkedList<>();
    static Scanner scanner = new Scanner(System.in);
    static int personasAtendidas = 0;
    
    public static void main(String[] args) {
        int opcion;
        
        // Agregar algunas personas de ejemplo
        fila.addLast(new Persona("Ana Martínez", "Consulta", 2));
        fila.addLast(new Persona("Carlos López", "Trámite", 3));
        fila.addLast(new Persona("Rosa García", "Urgencia", 1));
        
        do {
            mostrarMenu();
            System.out.print("Selecciona una opción: ");
            opcion = scanner.nextInt();
            scanner.nextLine();
            
            switch (opcion) {
                case 1:
                    agregarPersona();
                    break;
                case 2:
                    verFila();
                    break;
                case 3:
                    atenderPersona();
                    break;
                case 4:
                    verEstadisticas();
                    break;
                case 5:
                    simularAtencion();
                    break;
                case 6:
                    verPersonasAtendidas();
                    break;
                case 7:
                    System.out.println("\n¡Gracias por usar el sistema!");
                    break;
                default:
                    System.out.println("\n❌ Opción inválida.\n");
            }
        } while (opcion != 7);
        
        scanner.close();
    }
    
    static void mostrarMenu() {
        System.out.println("\n========== SISTEMA DE FILA DE ATENCIÓN ==========");
        System.out.println("1. Agregar persona a la fila");
        System.out.println("2. Ver fila de espera");
        System.out.println("3. Atender siguiente persona");
        System.out.println("4. Ver estadísticas");
        System.out.println("5. Simular atención rápida (5 personas)");
        System.out.println("6. Ver personas ya atendidas");
        System.out.println("7. SALIR");
        System.out.println("================================================");
    }
    
    // Agregar una persona a la fila
    static void agregarPersona() {
        System.out.print("\nIngresa el nombre: ");
        String nombre = scanner.nextLine();
        
        System.out.print("Ingresa el motivo: ");
        String motivo = scanner.nextLine();
        
        System.out.println("\nNivel de prioridad:");
        System.out.println("  1 - Urgencia (Prioridad ALTA)");
        System.out.println("  2 - Normal (Prioridad NORMAL)");
        System.out.println("  3 - Consulta (Prioridad BAJA)");
        System.out.print("Selecciona (1-3): ");
        int prioridad = scanner.nextInt();
        scanner.nextLine();
        
        if (prioridad < 1 || prioridad > 3) {
            System.out.println("❌ Prioridad inválida. Se asignó NORMAL.");
            prioridad = 2;
        }
        
        Persona nueva = new Persona(nombre, motivo, prioridad);
        
        // Insertar respetando prioridades
        if (prioridad == 1) {
            // Alta prioridad: ir al principio de otras altas
            int posicion = 0;
            for (int i = 0; i < fila.size(); i++) {
                if (fila.get(i).prioridad < 1) break;
                posicion = i + 1;
            }
            fila.add(posicion, nueva);
        } else {
            // Normal o baja: ir al final
            fila.addLast(nueva);
        }
        
        System.out.println("✓ " + nombre + " agregado a la fila.");
        System.out.println("  Posición en la fila: #" + (obtenerPosicion(nueva) + 1));
    }
    
    // Ver fila de espera
    static void verFila() {
        if (fila.isEmpty()) {
            System.out.println("\n✓ La fila está vacía. No hay personas esperando.");
            return;
        }
        
        System.out.println("\n========== FILA DE ESPERA ==========");
        System.out.println("Total personas en fila: " + fila.size() + "\n");
        
        for (int i = 0; i < fila.size(); i++) {
            String marca = (i == 0) ? "👤 [ATENDIENDO] " : "  [" + (i + 1) + "] ";
            System.out.println(marca + fila.get(i));
        }
    }
    
    // Atender a la siguiente persona
    static void atenderPersona() {
        if (fila.isEmpty()) {
            System.out.println("\n✓ No hay personas en la fila.");
            return;
        }
        
        Persona persona = fila.removeFirst();
        long tiempoEspera = persona.getTiempoEspera();
        personasAtendidas++;
        
        System.out.println("\n========== ATENDIENDO ==========");
        System.out.println("✓ " + persona.nombre);
        System.out.println("  ID: #" + persona.id);
        System.out.println("  Motivo: " + persona.motivo);
        System.out.println("  Prioridad: " + persona.getPrioridadTexto());
        System.out.println("  Tiempo de espera: " + tiempoEspera + " minutos");
        
        System.out.print("\n¿Cuánto tiempo tardará la atención? (segundos): ");
        int segundos = scanner.nextInt();
        scanner.nextLine();
        
        System.out.println("⏳ Atendiendo a " + persona.nombre + "...");
        
        // Simular tiempo de atención
        try {
            Thread.sleep(Math.min(segundos * 100, 3000)); // Máximo 3 segundos
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        
        persona.getTiempoEspera(); // Actualizar tiempo
        atendidas.addLast(persona);
        System.out.println("✓ Atención completada.");
        System.out.println("  Personas en la fila: " + fila.size());
    }
    
    // Ver estadísticas
    static void verEstadisticas() {
        System.out.println("\n========== ESTADÍSTICAS ==========");
        System.out.println("Personas en fila: " + fila.size());
        System.out.println("Personas atendidas: " + personasAtendidas);
        System.out.println("Total procesadas: " + (fila.size() + personasAtendidas));
        
        if (!fila.isEmpty()) {
            System.out.println("\n✓ Próxima persona a atender:");
            System.out.println("  " + fila.getFirst());
            
            System.out.println("\n✓ Última persona en fila:");
            System.out.println("  " + fila.getLast());
            
            // Estadísticas por prioridad
            int alta = 0, normal = 0, baja = 0;
            long tiempoPromedioEspera = 0;
            
            for (Persona p : fila) {
                if (p.prioridad == 1) alta++;
                else if (p.prioridad == 2) normal++;
                else baja++;
                tiempoPromedioEspera += p.getTiempoEspera();
            }
            
            System.out.println("\n✓ Distribución por prioridad:");
            System.out.println("  🔴 Urgencia: " + alta);
            System.out.println("  🟡 Normal: " + normal);
            System.out.println("  🟢 Baja: " + baja);
            
            if (fila.size() > 0) {
                System.out.printf("  Tiempo promedio espera: %.1f min\n", 
                    (double) tiempoPromedioEspera / fila.size());
            }
        }
    }
    
    // Simular atención rápida
    static void simularAtencion() {
        if (fila.isEmpty()) {
            System.out.println("\n✓ No hay personas en la fila para atender.");
            return;
        }
        
        int atender = Math.min(5, fila.size());
        System.out.println("\n⏳ Simulando atención a " + atender + " personas...\n");
        
        for (int i = 0; i < atender; i++) {
            if (!fila.isEmpty()) {
                Persona persona = fila.removeFirst();
                atendidas.addLast(persona);
                personasAtendidas++;
                System.out.println((i + 1) + ". ✓ " + persona.nombre + " - Atendido");
                
                try {
                    Thread.sleep(500);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        }
        
        System.out.println("\n✓ Simulación completada.");
        System.out.println("Personas en fila: " + fila.size());
        System.out.println("Personas atendidas: " + personasAtendidas);
    }
    
    // Ver personas ya atendidas
    static void verPersonasAtendidas() {
        if (atendidas.isEmpty()) {
            System.out.println("\n✓ No hay personas atendidas aún.");
            return;
        }
        
        System.out.println("\n========== PERSONAS ATENDIDAS ==========");
        System.out.println("Total: " + atendidas.size() + "\n");
        
        for (int i = 0; i < atendidas.size(); i++) {
            System.out.println((i + 1) + ". " + atendidas.get(i));
        }
    }
    
    // Método auxiliar para obtener posición de una persona
    static int obtenerPosicion(Persona persona) {
        for (int i = 0; i < fila.size(); i++) {
            if (fila.get(i).id == persona.id) {
                return i;
            }
        }
        return -1;
    }
}
